frontend/server/express.mjs

```diff
@@ -211,6 +270,231 @@ function startExpressServer() {
+      const response = {
+        has_config: true,
+        config: config,
+      };
+      res.status(200).send(response);
+    } else {
+      const response = { has_config: false };
+      res.status(200).send(response);
+    }
+  });
+
+  app.post("/launch-anvil-from-config", (req, res) => {
+    if (!electronApp.isPackaged) {
+      return res.sendStatus(200);
+    }
+
+    const anvilTimeoutPromise = new Promise((resolve, reject) => {
+      setTimeout(
+        () => {
+          reject(new Error("Anvil Timeout Error"));
+        },
+        3 * 60 * 1000,
+      );
+    });
+    if (anvilProcess !== null) {
+      anvilProcess.kill("SIGINT");
+      anvilProcess = null;
+      console.log("killed the process");
+    }
+    const anvilDir = path.join(process.resourcesPath, "server2");
+    const configPath = path.join(anvilDir, "config.json");
+    const configStr = fs.readFileSync(configPath);
+    const config = JSON.parse(configStr);
+
+    const context = config.KubeContext;
+    const kubeConfig = ["config", "use-context", context];
+    const kubeExec = spawn("kubectl", kubeConfig);
+
+    kubeExec.stderr.on("data", (data) => {
+      console.log(`stderr: ${data}`);
+      res
+        .status(500)
+        .send({ err: "CAN'T SET KUBECONTEXT", kubeErr: data.toString() });
+    });
+
+    const anvilFiles = fs.readdirSync(anvilDir);
+    const anvilExec = anvilFiles.filter((file) => file.startsWith("s2-"))[0];
+    const runAnvil = new Promise((resolve, reject) => {
+      anvilProcess = spawn(path.join(anvilDir, anvilExec), {
+        cwd: anvilDir,
+        detached: true,
+        stdio: ["ignore", "pipe", "pipe"], // Ignore stdin, use pipes for stdout and stderr
+      });
+
+      anvilProcess.stdout.on("data", (data) => {
+        console.log(`[server] stdout: ${data}`);
+
+        if (
+          data.toString().includes("[GIN-debug] Listening and serving HTTP on")
+        ) {
+          resolve();
+        }
+      });
+
+      anvilProcess.stderr.on("data", (data) => {
+        console.log(`[server] stderr: ${data}`);
+        if (
+          data
+            .toString()
+            .toLowerCase()
+            .includes("failed to fetch kubernetes resources;") ||
+          data
+            .toString()
+            .toLowerCase()
+            .includes("failed to get client config;") ||
+          data.toString().toLowerCase().includes("failed to install argo;")
+        ) {
+          reject(new Error(`Kubeservices not found: ${data.toString()}`));
+        }
+      });
+
+      anvilProcess.on("close", (code) => {});
+    });
+
+    const runAnvilPromise = Promise.race([anvilTimeoutPromise, runAnvil]);
+
+    runAnvilPromise
+      .then((response) => {
+        res.sendStatus(200);
+      })
+      .catch((err) => {
+        res
+          .status(500)
+          .send({ err: "Error while launching anvil", kubeErr: err.message });
+      });
+  });
+
+  app.post("/launch-anvil", (req, res) => {
```

Comment:
I’m not sure if I understand the difference between this endpoint and the one above it.

Reply:
One of them is basically starts the anvil runs from an existing config, whereas the other creates/re-creates the config.